src/SortingAlgorithms.jl

```diff
@@ -631,4 +683,308 @@ else
+next_page_B(pages::Pages) = Pages(pages.nextB, pages.currentNumber + 1, pages.nextA, pages.nextB + 1)
+
+Base.@propagate_inbounds function next_page!(pageLocations, pages, pagesize, lo, a)
+    if a > pages.nextA * pagesize + lo
+        pages = next_page_A(pages)
+    else
+        pages = next_page_B(pages)
+    end
+    pageLocations[pages.currentNumber] = pages.current
+    pages
+end
+
+Base.@propagate_inbounds function permute_pages!(f, v, pageLocations, page_offset, pagesize, page)
+    while f(page)
+        plc = pageLocations[page-3] # plc has data belonging to page
+        pageLocations[page-3] = page
+        copyto!(v, page_offset(page)+1, v, page_offset(plc)+1, pagesize)
+        page = plc
+    end
+    page
+end
+
+# merge v[lo:m] (A) and v[m+1:hi] (B) using buffer buf in O(sqrt(n)) space
+function paged_merge!(v::AbstractVector{T}, lo::Integer, m::Integer, hi::Integer, o::Ordering, buf::AbstractVector{T}, pageLocations::AbstractVector{<:Integer}) where T
+    @assert lo < m < hi
+    a = lo
+    b = m + 1
+    lenA = 1 + m - lo
+    lenB = hi - m
+
+    # regular merge if buffer is big enough
+    if lenA <= length(buf)
+        merge!(v, lo, m, hi, o, buf)
+        return
+    elseif lenB <= length(buf)
+        # TODO ?
+        # does not occur in balanced mergesort where length(A) <= length(B)
+        error("not implemented")
+        return
+    end
+
+    len = lenA + lenB
+    pagesize = isqrt(len)
+    nPages = len ÷ pagesize # a partial page at the end does not count
+    @assert length(buf) >= 3pagesize
+    @assert length(pageLocations) >= nPages - 3
+
+    @inline page_offset(page) = (page-1)*pagesize + lo - 1
+
+    @inbounds begin
+        ##################
+        # merge
+        ##################
+        # merge the first 3 pages into buf
+        a,b,_ = merge!((_,_,k) -> k<=3pagesize,buf,v,v,o,a,b,1)
+        # initialize variables for merging into pages
+        pages = Pages(-17, 0, 1, (m-lo) ÷ pagesize + 2) # first argument is unused
+        # more efficient loop while more than pagesize elements of A and B are remaining
+        while_condition1(offset) = (_,_,k) -> k <= offset + pagesize
+        while a < m-pagesize && b < hi-pagesize
+            pages = next_page!(pageLocations, pages, pagesize, lo, a)
+            offset = page_offset(pages.current)
+            a,b,_ = merge!(while_condition1(offset),v,v,v,o,a,b,offset+1)
+        end
+        # merge until either A or B is empty or the last page is reached
+        k, offset = nothing, nothing
+        while_condition2(offset) = (a,b,k) -> k <= offset + pagesize && a <= m && b <= hi
+        while a <= m && b <= hi && pages.currentNumber + 3 < nPages
+            pages = next_page!(pageLocations, pages, pagesize, lo, a)
+            offset = page_offset(pages.current)
+            a,b,k = merge!(while_condition2(offset),v,v,v,o,a,b,offset+1)
+        end
```

Comment:
These while loops and `while_condition`s feel strange to me; I wonder if there is a better approach.

Reply:
It was the best way to implement this I could think of. But of course I am open to suggestions for implrovement.